src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -82,35 +83,114 @@ private async Task InvokeInternal(HttpContext context)
                 var list = new List<KeyValuePair<string, string?>>(
                     request.Headers.Count + DefaultRequestFieldsMinusHeaders);
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestProtocol))
+                if (options.ModifyRequestLog != null)
                 {
-                    AddToList(list, nameof(request.Protocol), request.Protocol);
-                }
+                    var headerDictionary = new HeaderDictionary();
+                    var loggingContext = new HttpRequestLoggingContext(context, options, headerDictionary);
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestMethod))
-                {
-                    AddToList(list, nameof(request.Method), request.Method);
-                }
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestProtocol))
```

Comment:
InvokeInternal is too big. Extract these `if` groups to helper methods so we only see the main control flow here.